OE0-37: Satisfy OpenIMISActivityDefinition profile

malinowskikam opened the pull request:
Changes:
- Added `actvityDefinitionMapping` containing mappings needed for fhir-imis conversion
- Updated `activityDefinitionConverter` to satisfy current ActivityDefinition profile

[https://openimis.atlassian.net/browse/OE0-37](https://openimis.atlassian.net/browse/OE0-37)

dborowiecki commented on api_fhir_r4/converters/activityDefinitionConverter.py:
```diff
@@ -73,18 +80,48 @@ def get_imis_obj_by_fhir_reference(cls, reference, errors=None):
+    @classmethod
+    def build_imis_frequency(cls, imis_activity_definition, fhir_activity_definition, errors):
+        value = None
+        try:
```
I like the solution with the try, except, finally block for assigning values / adding information to the errors list. 
It's quite repetitive in every method. Do you think we could make it more generic, e.g. with creating a wrapper  implementing this logic in parent class and apply it for every build_imis_* method? 
It could also be reused in other converters for fhir -> IMIS mapping, and therefore improve error handling. @delcroip what do you think about it?

delcroip replied:
it sound possible https://stackoverflow.com/a/27500036/3533113

but it might be tricky because the assignment is on  "fhir_activity_definition.timingTiming.repeat.period" but you check if fhir_activity_definition have timingTiming attributes, I think a recursive function might be required

dborowiecki replied:
I'm not sure we need recursive approach. I think `hasattr(fhir_activity_definition, "timingTiming")` is only for checking if timing was provided in the payload. If it is there, we can assume that it's correctly formatted and contains repeat and period. However if there are issues with nested elements then message is added to the errors list.

malinowskikam replied:
That was the idea.

dragos-dobre replied:
This is a nice to have. https://openimis.atlassian.net/browse/OE0-54

dborowiecki reviewed (commented):
Looks good to me. @malinowskikam I've also left some comments, please refer to them

delcroip commented on api_fhir_r4/converters/activityDefinitionConverter.py:
```diff
@@ -73,18 +80,48 @@ def get_imis_obj_by_fhir_reference(cls, reference, errors=None):
-        cls.build_all_identifiers(identifiers, imis_activity_definition)
+        cls.build_fhir_uuid_identifier(identifiers, imis_activity_definition)
+        cls.build_fhir_code_identifier(identifiers, imis_activity_definition)
+        if reference_type == ReferenceConverterMixin.DB_ID_REFERENCE_TYPE:
```
then it could make sense to put it on the parent and override it in diagnosis, @dborowiecki  what do you think ?

dborowiecki replied:
I agree, by default we should use this approach and if for any reason conveners need customized solution, then they should override it.

delcroip commented on api_fhir_r4/converters/activityDefinitionConverter.py:
```diff
@@ -119,301 +156,247 @@ def build_fhir_title(cls, fhir_activity_definition, imis_activity_definition):
+            imp_coding = cls.build_fhir_venue_coding(Service.CARE_TYPE_IN_PATIENT)
+            codeable_concept.coding.append(imp_coding)
+
+        codeable_concept.text = " or ".join([coding.display for coding in codeable_concept.coding])
```
why not removing the coding if the coding.display is empty ? a complementary solution would be to use CODE if display is empty

dragos-dobre replied:
We need to define the code system used. useContext, parent item, cannot be code.

delcroip commented on api_fhir_r4/converters/activityDefinitionConverter.py:
```diff
@@ -119,301 +156,247 @@ def build_fhir_title(cls, fhir_activity_definition, imis_activity_definition):
-        codeable_concept.text = "Service Level"
+        codeable_concept.coding = []
+
+        male_flag = PatientCategoryMapping.imis_patient_category_flags["male"]
```
@dragos-dobre  I think we should no put any usecontext is it is open to all and have a specific "blocked" context if none is activated

dragos-dobre replied:
Note taken. The "blocked" context should be validated when adding Items and Services in openIMIS.

delcroip commented on api_fhir_r4/converters/activityDefinitionConverter.py:
```diff
@@ -119,301 +156,247 @@ def build_fhir_title(cls, fhir_activity_definition, imis_activity_definition):
         return codeable_concept
 
     @classmethod
-    def build_fhir_level(self, imis_activity_definition: Service):
```
@dragos-dobre , @malinowskikam informed me that there is not usecontext for the service level; personally I would use "jurisdiction" instead

dragos-dobre replied:
@delcroip I don't understand this comment. These values are removed and not defined in new profile.

dragos-dobre commented on api_fhir_r4/mapping/actvityDefinitionMapping.py:
```diff
@@ -0,0 +1,136 @@
+from medical.models import Service
```
Check the file name because is called actvityDefinitionMapping.py => missing an 'i'.

dragos-dobre commented on api_fhir_r4/mapping/actvityDefinitionMapping.py:
```diff
@@ -0,0 +1,136 @@
+    }
+
+
+class PatientCategoryMapping(object):
```
This mapping has also defined by @sniedzielski in medicationMapping.py. Having the same URI, it should be defined only once.

malinowskikam commented:
@dragos-dobre I moved PatientCategoryMapping from both `medicationMapping.py` and `activityDefinitionMapping.py` to `patientMapping.py`. I also fixed the typo, and changed `build_fhir_identifiers` to use generic implementation (which is the same).
